README.md
# [![Build Status](https://travis-ci.org/Miachol/configr.svg)](https://travis-ci.org/Miachol/configr) [![License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat)](https://en.wikipedia.org/wiki/MIT_License) [![CRAN](http://www.r-pkg.org/badges/version/configr)](https://cran.r-project.org/package=configr) [![Downloads](http://cranlogs.r-pkg.org/badges/configr?color=brightgreen)](http://www.r-pkg.org/pkg/configr) [![codecov](https://codecov.io/github/Miachol/configr/branch/master/graphs/badge.svg)](https://codecov.io/github/Miachol/configr) 

configr package
==============


The R package '[configr](https://github.com/Miachol/configr)' implements the
'[JSON](https://CRAN.R-project.org/package=jsonlite)', 
'[INI](https://CRAN.R-project.org/package=ini)', 
'[YAML](https://CRAN.R-project.org/package=yaml)', 
and '[TOML](https://CRAN.R-project.org/package=RcppTOML)' parser for R setting and writing of configuration file.

# Introduction 

The configuration file are necessary for many projects that will help us to manage and set project environment variables easily.

Configuration files, from INI/XML/JSON/YAML to TOML, readability and maneuverability have been improved too much in the past few years, and there are serveral parsers be created in R and other programming language. That have made us becomes more efficient, but, we need to remember the different functions for different format configuration file that sometimes we only just want to read it and regardless of that format. So, using single function to read or/and write most of configuration file are good way to reduce memory burden.


[configr](https://github.com/Miachol/configr) have done some work to relax us on configuration files that can be used to parse and generate JSON/INI/YAML/TOML format configuration file. The functionality of this package is similar to that of package '[config](https://CRAN.R-project.org/package=config)'.


## Configuration file format

### JSON
``` json
{   "default":{
        "debug":"{{debug}} {{debug2}}"
    },
    "comments":{
        "version":"0.2.0",
        "test_parse":"{{key:test_parse}} {{key:test_parse2}}",
        "test_parse2":"@>@ str_replace('{{key:test_parse}}', '2', '00')@<@ {{key:test_parse2}} {{debug2}}",
        "test_parse3":"{{key:test_parse}}",
        "test_parse4":"@>@ str_replace('{{key:test_parse2}}', '2', '00')@<@ @>@ str_replace('{{key:test_parse}}', '2', '00')@<@ {{key:test_parse2}} {{debug2}}"
    }
}
```
More infomation and example of JSON can be founded in [json.org](http://www.json.org/), [JSON Example](http://www.json.org/example.html) and [JSON-wikipedia](https://en.wikipedia.org/wiki/JSON). `{{key:key:value}}/{{key}}` can be parsed by parse.extra using `extra.list` and `other.config` parameters. `@>@ str_replace("123", "2", "1")@<@` can be parsed by `parse.extra` setting parameter `rcmd.parse` to `TRUE`. Example of that can be founded in this document tail.

### INI
``` ini
[default]
debug = {{debug}} {{debug2}}

[comments]
version = 0.2.0
test_parse = {{key:test_parse}} {{key:test_parse2}}
test_parse2 = @>@ str_replace('{{key:test_parse}}', '2', '00')@<@ {{key:test_parse2}} {{debug2}}
test_parse3 = {{key:test_parse}}
test_parse4 = @>@ str_replace('{{key:test_parse2}}', '2', '00')@<@ @>@ str_replace('{{key:test_parse}}', '2', '00')@<@ {{key:test_parse2}} {{debug2}}
```
More infomation and example of INI can be founded in [INI-wikipedia](https://en.wikipedia.org/wiki/INI_file).

### YAML
``` yaml
default:
  debug: '{{debug}} {{debug2}}'
comments:
  version: 0.2.0
  test_parse: '{{key:test_parse}} {{key:test_parse2}}'
  test_parse2: '@>@ str_replace(''{{key:test_parse}}'', ''2'', ''00'')@<@ {{key:test_parse2}}
    {{debug2}}'
  test_parse3: '{{key:test_parse}}'
  test_parse4: '@>@ str_replace(''{{key:test_parse2}}'', ''2'', ''00'')@<@ @>@ str_replace(''{{key:test_parse}}'',
    ''2'', ''00'')@<@ {{key:test_parse2}} {{debug2}}'
```
More infomation and example of YAML can be founded in [yaml.org](http://www.yaml.org/) and [YAML-wikipedia](https://en.wikipedia.org/wiki/YAML).

### TOML
``` toml
title = "TOML Example"

[default]
debug = true

[comments]
version = "0.1.0"
```
More infomation and example of TOML can be founded in and [toml-lang/toml](https://github.com/toml-lang/toml) and [TOML-wikipedia](https://en.wikipedia.org/wiki/TOML).

# Installation

## CRAN
``` r
#You can install this package directly from CRAN by running (from within R):
install.packages('configr')
```

## Github
``` r
# Install the cutting edge development version from GitHub:
# install.packages("devtools")
devtools::install_github("Miachol/configr")
```

## Zip/Tarball

1. Download the appropriate zip file or tar.gz file from Github
2. Unzip the file and change directories into the configr directory
3. Run `R CMD INSTALL pkg`

# Usage

``` r
#Get filepath of example configuration files
config.json <- system.file('extdata', 'config.json', package='configr')
config.ini <- system.file('extdata', 'config.ini', package='configr')
config.yaml <- system.file('extdata', 'config.yaml', package='configr')
config.toml <- system.file('extdata', 'config.toml', package='configr')
```
**is.json.file/is.ini.file/is.yaml.file/is.toml.file**

is.json.file/is.ini.file/is.yaml.file/is.toml.file can test configuration file type,
and more filetype will be supported in the future. If file.debug be setted to TRUE, it will show the warning infomation.
``` r
#Test file type of config file (No warning message)
is.json <- is.json.file(file = config.json)
is.ini <- is.ini.file(file = config.ini)
is.yaml <- is.yaml.file(file = config.yaml)
is.toml <- is.toml.file(file = config.toml)

#Test file type of config file (Debug, print warning message) 
is.json <- is.json.file(file = config.yaml, json.file.debug = T)
is.ini <- is.ini.file(file = config.json, ini.file.debug = T)
is.yaml <- is.yaml.file(file = config.toml, yaml.file.debug = T)
is.toml <- is.toml.file(file = config.yaml, toml.file.debug = T)
```

**get.config.type**

get.config.type can get configuration file type directly, it will return the json/ini/yaml/toml or FALSE for other format.
``` r
json <- get.config.type(file = config.json) 
ini <- get.config.type(file = config.ini) 
yaml <- get.config.type(file = config.yaml) 
toml <- get.config.type(file = config.toml) 
```

**read.config**

read.config can read a configuration file in R and as a list object that can pass parameter to inner read function (fromJSON/read.ini/yaml.load_file/parseToml) accordingly.
``` r
#Read in R as a list (JSON/INI/YAML/TOML be suported)
#fromJSON/read.ini/readLines/yaml.load  parameters can be automatch by parameter name (encoding .etc.)
json.list <- read.config(file = config.json)
ini.list <- read.config(file = config.ini)
yaml.list <- read.config(file = config.yaml)
toml.list <- read.config(file = config.toml) 
toml.list.parsed.1 <- read.config(file = config.toml, extra.list = list(debug = "TRUE")) 
other.config <- system.file('extdata', 'config.other.yaml', package='configr')
toml.list.parsed.2 <- read.config(file = config.toml, other.config = other.config) 
```

**eval.config**

eval.config return a value or a list object containing the file path, config group, filetype as the attribute.
``` r
#Get the same obj with config package, only get the 'default or R_CONFIG_ACTIVE config sets' in config.cfg or R_CONFIGFILE_ACTIVE
config.json.obj <- eval.config(file = config.json)
config.ini.obj <- eval.config(file = config.ini)
config.yaml.obj <- eval.config(file = config.yaml)
config.toml.obj <- eval.config(file = config.toml)
```

**eval.config.groups**

eval.config.groups can get all top groups in configuration file. 
``` r
json.groups <- eval.config.groups(file = config.json)
ini.groups <- eval.config.groups(file = config.ini)
yaml.groups <- eval.config.groups(file = config.yaml)
toml.groups <- eval.config.groups(file = config.toml)
```

**eval.config.merge**

eval.config.merge can read mulitple selected config groups as a list
(Default to read all groups, you can set parameter 'groups' to read whatever you want groups).
``` r
json.config.all <- eval.config.merge(file = config.json)
ini.config.all <- eval.config.merge(file = config.ini)
yaml.config.all <- eval.config.merge(file = config.yaml)
toml.config.all <- eval.config.merge(file = config.toml)
```

**convert.config**

convert.config can covert JSON/INI/YAML/TOML format configuration file to JSON/INI/YAML format configuration file 
``` r
convert.config(file = config.yaml, out.file = tempfile(, fileext = ".json"), convert.to = "JSON")
```

**write.config**

write.config as the only write function, JSON/INI/YAML be supported now, the TOML will be done soon. Of course, 
toJSON/write.ini/as.yaml can get the parameters from write.config
``` r
#Write config in file (JSON/INI/YAML be supported now, TOML be included TODO list, 
list.test <- list(a=c(123,456))
out.fn <- sprintf("%s/test.json", tempdir())
write.config(config.dat = list.test, file.path = out.fn, write.type = "json")

# toJSON/write.ini/as.yaml parameters can be automatch by parameter name (indent/encoding .etc.)
write.config(config.dat = list.test, file.path = out.fn, write.type = "json", indent = 2)
out.fn <- sprintf("%s/test.yaml", tempdir())
write.config(config.dat = list.test, file.path = out.fn, write.type = "yaml")
write.config(config.dat = list.test, file.path = out.fn, write.type = "yaml", indent = 4)

out.fn <- sprintf("%s/test.ini", tempdir())
write.config(config.dat = list.test, file.path = out.fn, write.type = "ini")
```

**parse.extra**

parse.extra default be used by read.config, eval.config and others by parameters "extra.list", "other.config", "rcmd.parse".
``` r
config.1 <- read.config(file = config.json)
other.config <- system.file('extdata', 'config.other.yaml', package='configr')
config.2 <- read.config(file = config.json, extra.list = list(debug = "self", debug2 = "self2"))
config.3 <- read.config(file = config.json, extra.list = list(debug = "self", debug2 = "self2"), other.config = other.config)
# The followed two line command will return the same value
config.4 <- read.config(file = config.json, extra.list = list(debug = "self", debug2 = "self2"), other.config = other.config, rcmd.parse = T)
config.5 <- parse.extra(config.1, extra.list = list(debug = "self", debug2 = "self2"), other.config = other.config, rcmd.parse = T)

```


ChangeLog
2017-03-23 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* %'% in configuration when rcmd.parse=TRUE, can be parsed to double
		quote that can to avoid double quote escape which can raise error

2017-03-17 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>
		
		* 0.2.1 release
		* '@>@ R statement @<@' in config can be paresed in parse.extra(rcmd.parse =
		TRUE)
		* Fix mulitple var parse in parse.R parse.config

2017-03-09 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* 0.2.0 release
		* Add parse.extra example in read.config and eval.config

2017-03-07 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>
		
		* Add parse.extra function in R/parse.R
		* Import stringr
		* Change tests output format

2017-02-14 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* Use RcppTOML 0.1.1

2017-02-04 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* Cancel import config package

2017-01-29 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* Cancel import stringr package

2017-01-22 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>
		
		* Cancel import futile.logger package

2017-01-20 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* Add is.configfile.active/is.config.active in R/active.R
		* Change the default paramerter of configuration file to
		Sys.getenv("R_CONFIGFILE_ACTIVE", "config.cfg")

2017-01-19 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>
		
		* json.file.debug/ini.file.debug/yaml.file.debug/toml.file.debug be 
		added in is.*.file function family that can show the error message
		of the configuration file, which can be used in 'configr' most other function
		* Add config.error.json/config.error.yaml/config.error.ini/config.error.toml
		in inst/extdata as the test file that all of them have be modified as
		non-standard format(JSON/INI/YAML/TOML) configuration files
		* Add additional tests for code coverage
		* Rename inst/config.yml to inst/config.yaml

2017-01-17 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* Add TOML parse (RcppTOML)
		* Add convert.config function in R/convert.R
		* Use '...' as the parameter rather than 'encoding' only


R/merge.R
#' Merge config parameter sets by groups.
#'
#' @param file File name of configuration file to read from. Defaults to the value of
#' the 'R_CONFIGFILE_ACTIVE' environment variable ('config.cfg' if the
#' variable does not exist and JSON/INI/YAML/TOML format only)
#' @param ... Arguments for \code{\link{get.config.type}}, 
#' \code{\link{eval.config.groups}}, \code{\link{eval.config}}
#' @param groups Need be merged parameter sets, eg. groups=c('default', 'version'), will 
#' default to all of config groups
#' @seealso
#' \code{\link{eval.config.groups}} which only get all of the mainly parameter sets name in config file, 
#' \code{\link{read.config}} which only read from a config as a list,
#' \code{\link{eval.config}} which only read one groups as config obj or a value from config file.
#' @return A list or logical FALSE indicating that is not standard JSON/INI/YAML/TOML format file
#' @export
#' @examples
#' config.json <- system.file('extdata', 'config.json', package='configr')
#' config.ini <- system.file('extdata', 'config.ini', package='configr')
#' config.yaml <- system.file('extdata', 'config.yaml', package='configr')
#' config.toml <- system.file('extdata', 'config.toml', package='configr')
#' eval.config.merge(config.json)
#' eval.config.merge(config.ini)
#' eval.config.merge(config.yaml)
#' eval.config.merge(config.toml)
eval.config.merge <- function(file = Sys.getenv("R_CONFIGFILE_ACTIVE", "config.cfg"), 
  groups = NULL, ...) {
  config.dat <- list()
  groups.all <- eval.config.groups(file = file, ...)
  config.type <- get.config.type(file = file, ...)
  if (is.logical(groups.all) && groups.all == FALSE) {
    return(FALSE)
  }
  if (is.null(groups)) {
    for (i in groups.all) {
      config.tmp <- eval.config(file = file, config = i, ...)
      if (!is.list(config.tmp)) {
        config.tmp <- as.list(config.tmp)
        names(config.tmp) <- i
      }
      config.dat <- config.list.merge(list.left = config.dat, list.right = config.tmp)
    }
    attr(config.dat, "config") <- groups.all
  } else {
    groups <- groups[groups %in% groups.all]
    for (i in groups) {
      config.tmp <- eval.config(file = file, config = i, ...)
      if (!is.list(config.tmp)) {
        config.tmp <- as.list(config.tmp)
        names(config.tmp) <- i
      }
      config.dat <- config.list.merge(list.left = config.dat, list.right = config.tmp)
      attr(config.dat, "config") <- groups
    }
  }
  attr(config.dat, "configtype") <- config.type
  attr(config.dat, "file") <- file
  return(config.dat)
}

#' Merge list file (From config package), list.right will overwrite the element also existed in list.left
#'
#' @param list.left One list be merged left
#' @param list.right One list be merged right 
#' @seealso
#' \code{\link[config]{merge}} call in this function
#' @return A list 
#' @export
#' @examples
#' config.json <- system.file('extdata', 'config.json', package='configr')
#' list.left <- list()
#' list.right <- eval.config(file = config.json)
#' config.list.merge(list.left, list.right)
#' list.left <- list(a=c(123,456))
#' list.right <- list(a=c(4,5,6)) 
#' config.list.merge(list.left, list.right)
config.list.merge <- function(list.left = list(), list.right = list()) {
  if (!is.list(list.left) | !is.list(list.right)) {
    warning("list.left or list.left must be list type!")
  }
  list.merge(list.left, list.right)
}
